geode-redis/src/distributedTest/java/org/apache/geode/redis/OutOfMemoryDUnitTest.java

```diff
@@ -0,0 +1,188 @@
+  private static MemberVM server2;
+
+  @BeforeClass
+  public static void classSetup() {
+    Properties serverProperties;
+    MemberVM locator;
+    int redisServerPort1;
+    int redisServerPort2;
+
+    IgnoredException.addIgnoredException(expectedEx);
+    serverProperties = new Properties();
+
+    locator = clusterStartUp.startLocatorVM(0);
+    server1 = clusterStartUp.startRedisVM(1, serverProperties, locator.getPort());
+    server2 = clusterStartUp.startRedisVM(2, serverProperties, locator.getPort());
+
+    server1.getVM().invoke(() -> RedisClusterStartupRule.getCache().getResourceManager()
+        .setCriticalHeapPercentage(5.0F));
+    server2.getVM().invoke(() -> RedisClusterStartupRule.getCache().getResourceManager()
+        .setCriticalHeapPercentage(5.0F));
+
+    redisServerPort1 = clusterStartUp.getRedisPort(1);
+    redisServerPort2 = clusterStartUp.getRedisPort(2);
+
+    jedis1 = new Jedis(LOCAL_HOST, redisServerPort1, JEDIS_TIMEOUT);
+    jedis2 = new Jedis(LOCAL_HOST, redisServerPort2, JEDIS_TIMEOUT);
+  }
+
+  @Before
+  public void testSetup() {
+    jedis1.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis1.disconnect();
+    jedis2.disconnect();
+
+    server1.stop();
+    server2.stop();
+  }
+
+  @Test
+  public void shouldReturnOOMError_forWriteOperations_whenThresholdReached() {
+    IgnoredException.addIgnoredException(expectedEx);
+    IgnoredException.addIgnoredException("LowMemoryException");
+
+    fillMemory(jedis1, false);
+
+    assertThatThrownBy(() -> jedis2.set("oneMoreKey", makeLongStringValue(2 * LARGE_VALUE_SIZE)))
+        .hasMessageContaining("OOM");
+  }
+
+  @Test
+  public void shouldAllowDeleteOperations_afterThresholdReached() {
+    IgnoredException.addIgnoredException(expectedEx);
+    IgnoredException.addIgnoredException("LowMemoryException");
+
+    fillMemory(jedis1, false);
+
+    assertThatNoException().isThrownBy(() -> jedis2.del(FILLER_KEY + 1));
+  }
+
+  @Test
+  public void shouldAllowExpiration_afterThresholdReached() {
+    IgnoredException.addIgnoredException(expectedEx);
+    IgnoredException.addIgnoredException("LowMemoryException");
+
+    fillMemory(jedis1, true);
+
+    GeodeAwaitility.await().until(() -> jedis2.ttl(FILLER_KEY + 1) == -2);
+  }
+
+  // TODO: test that write operations become allowed after memory has dropped
+  // below critical levels. Difficult to do right now because of vagaries of the
+  // Java garbage collector.
+
+  private void fillMemory(Jedis jedis, boolean withExpiration) {
+    String valueString;
+    int valueSize = LARGE_VALUE_SIZE;
+
+    forceGC(); // Helps ensure we really do fill all available memory
+
+    while (valueSize > 1) {
+      valueString = makeLongStringValue(LARGE_VALUE_SIZE);
+      addMultipleKeys(jedis, valueString, withExpiration);
+      valueSize /= 2;
+    }
+  }
+
+  private void addMultipleKeys(Jedis jedis, String valueString, boolean withExpiration) {
+    int i = 0;
```

Comment:
For-loop with `catchThrowable` is another option here. `catchThrowable` is using a lambda and lambdas cannot use a non-final int so... I used an `AtomicInteger` which is final but the value it contains is mutable.
```
import static org.assertj.core.api.Assertions.catchThrowable;

// count is final because it is never reassigned
AtomicInteger count = new AtomicInteger();

Throwable thrown = catchThrowable(() -> {
  for (count.set(0); count.get() < MAX_ITERATION_COUNT; count.incrementAndGet()) {
    setRedisKeyAndValue(jedis, withExpiration, valueString, count.get());
  }
});

assertThat(thrown)
    .isInstanceOf(Exception.class)
    .hasMessageContaining("OOM command not allowed");

assertThat(count.get()).isLessThan(MAX_ITERATION_COUNT);
```

Reply:
I like it!